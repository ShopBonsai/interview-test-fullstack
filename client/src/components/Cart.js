import React, { useContext, Fragment } from 'react';
import Context from '../context';
import './cartContainer.css';

const Cart = () => {
  const { state: { productsInCart }, dispatch } = useContext(Context);
  if (productsInCart.length < 1) return <span></span>
  return (
    <div className="cart-container">
      <p><strong>Your ðŸ›’ with your beautiful products ðŸ˜‰</strong></p>
      <hr />
      {productsInCart.map(({ id, name, image }) => (
        <Fragment key={id}>
          <div className="cart-product-container" key={id}>
            <img src={image} />
            <div className="card-product-description">
              <p>{name}</p>
              <button id={id} onClick={({ target: { id }}) => dispatch({ type: "REMOVE_PRODUCT_FROM_CART", payload: { id } })} className="remove">Remove item</button>
            </div>
          </div>
          <hr />
        </Fragment>
      ))}
      <button onClick={() => alert('boom, finished.')}>Finish Purchase</button>
    </div>
  )
}

export default Cart;