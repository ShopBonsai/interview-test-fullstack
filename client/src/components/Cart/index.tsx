import React, { Fragment, useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Button, Card, CardText } from 'reactstrap';
import { RootState } from '../../interfaces';
import { removeFromCart, checkoutCart, toggleCart } from '../../actions';
import { toFixedNumber } from '../../utils/utils';
import CartItem from '../CartItem';

const Cart: React.FC = () => {
  // would probably handle different notifications in it's own reducer with more time
  const [notification, setNotification] = useState('');
  const dispatch = useDispatch();
  const cart = useSelector((state: RootState) => state.shop.cart.current);
  const isCartOpen = useSelector((state: RootState) => state.shop.cart.isOpen);
  const itemCount = useSelector((state: RootState) => state.shop.cart.itemCount);
  const subTotal = useSelector((state: RootState) => state.shop.cart.subTotal);

  useEffect(() => {
    if (!isCartOpen && notification) {
      setTimeout(() => {
        setNotification('');
      }, 1000);
    }
  }, [isCartOpen]);

  const removeItemFromCart = (id: string): void => {
    dispatch(removeFromCart(id));
  };

  const finalizeCheckout = async () => {
    const response = await fetch('/api/checkout', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        userId: 'guest',
        subTotal,
        itemCount,
        cart,
      }),
    });

    const { message } = await response.json();
    if (message) {
      dispatch(checkoutCart());
      setNotification(message);
    }
  };

  return (
    <div className={`cart ${isCartOpen && 'show'}`}>
      {notification ? (
        <Card body outline color="success" className="cart-item">
          <CardText className="text-success">{notification}</CardText>
          <Button color="success" onClick={() => dispatch(toggleCart())}>
            Back To Products
          </Button>
        </Card>
      ) : (
        <Fragment>
          <h3 className="cart-heading">Cart subtotal: ${subTotal}</h3>
          {cart.map((item) => {
            const { id, image, name, price, selectedQuantity } = item;
            return (
              <CartItem
                key={id}
                id={id}
                image={image}
                name={name}
                price={toFixedNumber(price * selectedQuantity)}
                selectedQuantity={selectedQuantity}
                removeItemFromCart={removeItemFromCart}
              />
            );
          })}
          <Button
            className="checkout"
            color="success"
            size="lg"
            block
            disabled={!cart.length}
            onClick={finalizeCheckout}
          >
            Checkout
          </Button>
        </Fragment>
      )}
    </div>
  );
};

export default Cart;
