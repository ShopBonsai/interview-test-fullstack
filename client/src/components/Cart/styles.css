.cart-item {
  display: flex;
  margin-top: 10px;
}

.cart-item-info {
  padding: 0 10px;
}

.cart-item-name {
  font-size: 1rem;
  color: grey;
}

.cart-title {
  font-size: 1.25rem;
  font-weight: 600;
  color: darkslategrey;
  margin-bottom: 10px;
}

.cart {
  padding: 10px;
}

.cart-no-items {
  font-size: 0.75rem;
}

.cart-count-info {
  font-size: 0.75rem;
}

.checkout-button-wrapper {
  display: flex;
  justify-content: center;
  margin-top: 20px;
}
