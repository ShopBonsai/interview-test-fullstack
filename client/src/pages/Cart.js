import React, {useState, useEffect} from 'react';
import { useSelector, useDispatch } from "react-redux";
import { Table, Container, Row, Col, Button } from 'reactstrap';
import QuantityIndicators from '../components/QuantityIndicators';
import { increasetItemQty, decreaseItemQty } from "../actions/cartAction";
import {useHistory} from "react-router-dom";

export default function Cart(props) {
  const dispatch= useDispatch();
  const history = useHistory();
  const [productTotal, setProductTotal] = useState(0);
  const cartItems = useSelector(state => state.cart.items);

  const handleQtyIncrease = (item) =>{
    dispatch(increasetItemQty(item));
  };

  const handleQtyDecrease = (item) =>{
    dispatch(decreaseItemQty(item));
  };

  const calculateProductTotal = (qty, price) =>{
    const total = price * qty;
    return total.toFixed(2);
  };

  const completeCheckout = () =>{
    return history.push("/checkout");
  };

  const calculateCartTotal = (items) =>{
    const totalCost = Object.values(items).reduce((total, {price, qty}) =>{
      return (total + (price* qty))
    }, 0);
    return totalCost;
  };

  return (
    <Container>
      <Row>
        <Col sm="12" md={{size: 6, offset: 3}}>
          <Table hover size="sm">
            <thead>
              <tr>
                <th>#</th>
                <th>Product Name</th>
                <th>Quantity</th>
                <th>Price</th>
              </tr>
            </thead>
            <tbody>
              {
                cartItems && Object.values(cartItems).map((item, idx) =>{
                  return (
                    <tr key={item.id}>
                      <th scope="row">{idx + 1}</th>
                      <td>{item.name}</td>
                      <td>
                        <QuantityIndicators
                          product={item}
                          increaseQty={handleQtyIncrease}
                          cartItem={item}
                          decreaseQty={handleQtyDecrease}
                        />
                      </td>
                      <td>${calculateProductTotal(item.price, item.qty)}</td>
                    </tr>
                  )
                })
              }
              <tr>
                <th colSpan="3">Cart Total: </th>
                <td colSpan="1"><b>${calculateCartTotal(cartItems).toFixed(2)}</b></td>
              </tr>
            </tbody>
          </Table>
          <div className="checkout">
            <Button 
              color="secondary" 
              block
              onClick={completeCheckout}
              disabled={cartItems && Object.values(cartItems).length < 1}
            >Checkout</Button>
          </div>
        </Col>
      </Row>
    </Container>
  )
};