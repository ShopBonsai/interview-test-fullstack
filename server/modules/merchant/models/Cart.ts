import mongoose from 'mongoose';

const Schema = mongoose.Schema;

export const CartSchema = new Schema({
    userId: String,
    items: [
        {
            id: String,
            name: String,
            image: String,
            price: Number,
            quantity: Number,
        },
    ],
});
