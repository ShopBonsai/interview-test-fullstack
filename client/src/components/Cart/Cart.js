import React from "react";
import { cartItemsVar } from "../../cache";
import { useReactiveVar } from "@apollo/client";
import { Media, Button } from "reactstrap";
import "./styles.css";

const CartItem = (props) => {
  const { id, image, name, count } = props.item;
  const handleRemove = () => {
    const cartItems = { ...cartItemsVar() };
    delete cartItems[id];
    cartItemsVar(cartItems);
  };

  const CartItem = (
    <div className="cart-item">
      <div>
        <Media
          object
          src={image}
          alt="Product image cap"
          width="90"
          height="100"
        />
      </div>
      <div className="cart-item-info ">
        <div className="cart-item-name">{name}</div>
        <div className="cart-count-info">
          <span>Count: </span>
          <span>{count}</span>
        </div>
        <Button color="danger" size="sm" onClick={handleRemove}>
          Remove
        </Button>
      </div>
    </div>
  );

  return CartItem;
};
const Cart = () => {
  const cartItems = useReactiveVar(cartItemsVar);
  const cartItemsArray = Object.values(cartItems);
  return (
    <div className="cart">
      <div className="cart-title">My Cart</div>
      {cartItemsArray.length === 0 ? (
        <p className="cart-no-items">No items in your cart</p>
      ) : (
        <div>
          {cartItemsArray.map((product) => (
            <CartItem key={product.id} item={product} />
          ))}
          <div className="checkout-button-wrapper">
            {/* This button is a placeholder */}
            <Button color="success">Checkout</Button>
          </div>
        </div>
      )}
    </div>
  );
};

export default Cart;
