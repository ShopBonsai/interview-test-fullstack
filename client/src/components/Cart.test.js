import React from 'react';
import renderer from 'react-test-renderer';
import Context from '../context';
import Cart from './Cart';

describe('Cart', () => {
  it('should render correctly', () => {
    expect(
      renderer.create(
        <Context.Provider value={{
          dispatch: jest.fn(),
          state: {
            productsInCart: [{ id: '1', name: 'test' }]
          },
        }}>
          <Cart />
        </Context.Provider>
      )
    ).toMatchSnapshot();
  });
});