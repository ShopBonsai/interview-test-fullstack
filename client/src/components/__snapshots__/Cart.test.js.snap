// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should render correctly 1`] = `
<div
  className="cart-container"
>
  <p>
    <strong>
      Your ðŸ›’ with your beautiful products ðŸ˜‰
    </strong>
  </p>
  <hr />
  <div
    className="cart-product-container"
  >
    <img />
    <div
      className="card-product-description"
    >
      <p>
        test
      </p>
      <button
        className="remove"
        id="1"
        onClick={[Function]}
      >
        Remove item
      </button>
    </div>
  </div>
  <hr />
  <button
    onClick={[Function]}
  >
    Finish Purchase
  </button>
</div>
`;
